### PR TITLE
fused layernorm

### PR DESCRIPTION
**Pull request by yang**
- Add simple util for timings
- Add fused layernorm kernel from Megatron

### CODE REVIEW

File: megatron/core/utils.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+from functools import reduce
```
Comment (Quentin-Anthony) [inline megatron/core/utils.py:1]: We don't have `megatron/core`, so this code needs added to `megatron/utils.py` along with updated imports

File: megatron/devutil.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    >>> monitor_method_wall_times(metric, layernorm, 'forward')
+    """
+    oldmeth = obj[methodname]
+
+    def newmeth(*args, **kw):
+        start = time.time()
+        try:
+            return oldmeth(*args, **kw)
+        finally:
+            end = time.time()
```
Comment (Quentin-Anthony) [inline megatron/devutil.py:31]: Using time.time() is inaccurate for CUDA code, especially distributed code. Use torch cuda events like in https://github.com/EleutherAI/cookbook/tree/main/benchmarks/communication

File: megatron/model/norms.py [hunk 3]
```diff
@@ -13,8 +13,8 @@
 # limitations under the License.
 
 import torch
-from torch.nn import LayerNorm as LayerNorm
-
+# from torch.nn import LayerNorm as LayerNorm
```
Comment (Quentin-Anthony) [inline megatron/model/norms.py:16]: We need to add a config arg for the fused layernorm, and turn it on by default. People may want to turn it off. We should treat the layernorm kernel the same as our other fused kernels: https://github.com/EleutherAI/gpt-neox/blob/98716ebb4818498fe118ddb07c62bfa94a48da2a/megatron/neox_arguments/neox_args.py#L243-L261

File: megatron/neox_arguments/neox_args.py [hunk 4]
```diff
@@ -125,6 +125,11 @@ class NeoXArgsModel(NeoXArgsTemplate):
     Normalization layer to use. Choose from "layernorm", "rmsnorm", "scalenorm".
     """
 
+    layernorm_fusion: bool = True
```
Comment (Quentin-Anthony) [inline megatron/neox_arguments/neox_args.py:128]: Can we do the same config line addition that was done for fused rope? I suspect users will want to tune this knob.

Comment (Quentin-Anthony) [reply megatron/neox_arguments/neox_args.py:128]: LGTM otherwise!

**Comment by jahatef** [comment]
Howdy,

Upon testing, we found no speedup and eventual slowdown when training a 1-3B model for 320 iterations. You can see my test [here](https://wandb.ai/jacobhatef/test_fused?workspace=user-hatef-4).

**Comment by StellaAthena** [comment]
> Howdy,
> 
> Upon testing, we found no speedup and eventual slowdown when training a 1-3B model for 320 iterations. You can see my test [here](https://wandb.ai/jacobhatef/test_fused?workspace=user-hatef-4).

It looks like the baseline the one with anomalous behavior here. Not to be an ass, but did you double and triple check by running it multiple times that something weird didn't just happen by magic to cause the baseline to speed up?

**Comment by jahatef** [comment]
@StellaAthena I agree, so I ran some extra tests, still available from the above link. I see no difference between fused layer norm being on/off within the variance of the runs.

**Comment by Quentin-Anthony** [comment]
I looked over these results with @jahatef and agree it falls within variance. I'm going to run some CUDA profiling to check if data movement is decreased with the layernorm kernel, and will merge if so. This would be on the grounds that as newer GPUs spend progressively less time on GEMMs, data movement becomes more critical.

**Comment by Quentin-Anthony** [comment]
I see meaningful decrease in data movement and confirmed the preservation of accuracy. Merging. Thanks a ton for this @yang

**Review by Quentin-Anthony — APPROVED** [review]

**Review by Quentin-Anthony — APPROVED** [review]
